Toolkit/ArcGISToolkit/TemplatePickerViewController.swift

```diff
@@ -0,0 +1,296 @@
+        
+        // when the user taps on a feature type
+        
+        // first get the selected object
+        let selectedFeatureTemplateInfo = infoForIndexPath(indexPath)!
+        
+        // If the search controller is still active, the delegate will not be
+        // able to dismiss us, if desired.
+        // Note: we can't do this before the above, or else the object we pull
+        // out of the datasource will be incorrect
+        navigationItem.searchController?.isActive = false
+        
+        delegate?.templatePickerViewControllerDidSelectTemplate(self, featureTemplateInfo: selectedFeatureTemplateInfo)
+    }
+    
+    override public func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        
+        guard tables.count > 0 else {
+            return 0
+        }
+        
+        let tableName = tables[section].tableName
+        let infos = self.currentDatasource[tableName]
+        return infos?.count ?? 0
+    }
+    
+    override public func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: cellReuseIdentifier, for: indexPath)
+        let info = infoForIndexPath(indexPath)!
+        cell.textLabel?.text = info.featureTemplate.name
+        cell.imageView?.image = info.swatch
+        return cell
+    }
+    
+    // MARK: go back, cancel methods
+    
+    @objc private func cancelAction(){
+        // If the search controller is still active, the delegate will not be
+        // able to dismiss us, if desired.
+        navigationItem.searchController?.isActive = false
+        delegate?.templatePickerViewControllerDidCancel(self)
+    }
+    
+    // MARK: IndexPath -> Info
+    
+    private func infoForIndexPath(_ indexPath: IndexPath) -> FeatureTemplateInfo?{
+        let tableName = tables[indexPath.section].tableName
+        let infos = self.currentDatasource[tableName]
+        return infos?[indexPath.row]
+    }
+    
+    private func indexPathForInfo(_ info: FeatureTemplateInfo) -> IndexPath?{
+        
+        let tableIndex = tables.index { $0.tableName == info.featureTable.tableName }
+        let infos = self.currentDatasource[info.featureTable.tableName]
+        let infoIndex = infos?.index { $0 === info }
+        
+        guard let section = tableIndex, let row = infoIndex else{
+            return nil
+        }
+        
+        return IndexPath(row: row, section: section)
+    }
+}
+
+extension TemplatePickerViewController: UISearchResultsUpdating {
+    public func updateSearchResults(for searchController: UISearchController) {
+        if let text = searchController.searchBar.text?.trimmingCharacters(in: .whitespaces),
+            !text.isEmpty {
+            isFiltering = true
+            DispatchQueue.global(qos: DispatchQoS.QoSClass.userInitiated).async {
```

Comment:
You're filtering on a separate queue? Interesting idea. But is the overhead worth it? Regardless, I see a few problems:

- Both `work` closures are strongly capturing `self`.
- What happens if `isFiltering` is `false` when the first `work` closure executes? Should it bail?
- What happens if `isFiltering` is `false` when the second `work` closure executes? Should it `guard` against that?
- What happens if `unfilteredInfos` has changed while dispatching to the main queue?

Reply:
great questions - I will try to address them